# Alternating list

The two most important parts of this section are that we will see how to define a function in Python, and we will get our first introduction to NumPy (which is the most important Python library for Math 9).  We will also see how to use an `if` statement in Python.

Our overall goal in this section is the following.

**Goal**: Write a function which takes input `n` and as output returns the length `n` list `[3,8,3,8,...]`.

## Writing a function that returns a constant list

<iframe width="560" height="315" src="https://www.youtube.com/embed/0i8ZhAUQ6Ik" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Before getting to our goal of defining a function that produces an alternating list, we will see how to define a function that returns a constant list of all 3s.  (Later in this course we will see a more elegant way to do this, using what is called "list comprehension", but here we will use a for loop.)

**Preliminary Goal**: Write a function which takes input `n` and as output returns the length `n` list `[3,3,3,...]`.

Here is a first attempt to define a function `f`.  The overall syntax is correct (especially the indentations), but it does not return anything, as we can see by trying to call `f(5)`.

In [1]:
def f(n):
    mylist = []
    for i in range(n):
        mylist.append(3)

In [2]:
f(5)

In the next definition we do correctly return something, but the indentation is wrong, the `return` statement should not be indented twice.  The second indentation is forcing the `return` statement to occur within the for loop, so we actually exit the function during the first step in the for loop.

In [3]:
def f(n):
    mylist = []
    for i in range(n):
        mylist.append(3)
        return mylist

In [4]:
f(5)

[3]

Here is a correct solution to our "preliminary goal" from above.

In [5]:
def f(n):
    mylist = []
    for i in range(n):
        mylist.append(3)
    return mylist

In [6]:
f(5)

[3, 3, 3, 3, 3]

In [7]:
f(6)

[3, 3, 3, 3, 3, 3]

In the above, you should think of `n` as getting assigned to 5 or to 6, but that assignment is only happening locally within the function itself.  The value of `n` is not accessible to us outside of the function.

In [8]:
n

NameError: name 'n' is not defined

Similarly, if we define `n` outside of the function (for example, `n = 100`), that does not have any influence on the `n` in the definition of the function.  Nor does the `n` inside of the function have any influence on the `n` that we have defined to be 100 outside of the function.

In [9]:
n = 100

In [10]:
f(4)

[3, 3, 3, 3]

In [11]:
n

100

## First version of the alternating function

<iframe width="560" height="315" src="https://www.youtube.com/embed/38tGfF_wqIE" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Now we will actually create the list that alternates between 3s and 8s.  Let's start by giving our function defined above a more descriptive name.  We will use `const` instead of `f`.

In [1]:
def const(n):
    mylist = []
    for i in range(n):
        mylist.append(3)
    return mylist

In [2]:
const(5)

[3, 3, 3, 3, 3]

We're definitely not done yet, but let's just get started by using our `const` function above.  It is definitely better to use the name `const` in the code, rather than copying the code from above.

In [3]:
def alt1(n):
    mylist = const(n)
    return mylist

In [4]:
alt1(4)

[3, 3, 3, 3]

Think of the next function as first making a constant list of all 3s, and then replacing all the odd-indexed elements by 8.  Using `range(1,n,2)` is a nice way to get access to those odd indices.

In [2]:
def alt1(n):
    mylist = const(n)
    for i in range(1,n,2):
        mylist[i] = 8
    return mylist

Let's check that the code works for both an even input as well as for an odd input.

In [3]:
alt1(10)

[3, 8, 3, 8, 3, 8, 3, 8, 3, 8]

In [7]:
alt1(11)

[3, 8, 3, 8, 3, 8, 3, 8, 3, 8, 3]

## Checking if the index is even or odd

<iframe width="560" height="315" src="https://www.youtube.com/embed/7iGgfOrsmRc" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Here we'll write a different version of the function, which will use an if statement to check if the index is even or odd.  Notice that the above `alt1` code is longer than it looks, since it is also using the `const` function that we wrote earlier.

Let's first see how to check if an integer is even or odd.  We use the `%` operator.  If you've seen modular arithmetic (for example in Math 13), `i%2` is like reducing $i$ modulo 2.  Another way to think about it is that the result is the remainder obtained when dividing $i$ by 2.

In [1]:
i = 7

In [2]:
i%2

1

In [3]:
6%2

0

We can check if an integer `i` is even by checking `i%2 == 0`.  We use an `if` statement, which works very similarly to in Matlab.

In [4]:
def alt2(n):
    mylist = []
    for i in range(n):
        if i%2 == 0:
            mylist.append(3)
        else:
            mylist.append(8)
    return mylist

In [5]:
alt2(5)

[3, 8, 3, 8, 3]

In [6]:
alt2(6)

[3, 8, 3, 8, 3, 8]

Here is a very similar version, where instead of `else` we use `elif`.  When using `elif`, you need to specify a condition.  These two versions do the same thing, because if the integer `i` is not even, it is always going to be odd.  You can have as many `elif` statements as you want, although in this case, `i%2 == 0` and `i%2 == 1` exhaust all the possibilities, so in this case, no later `elif` statements (nor `else` statements) would ever be reached.

In [4]:
def alt2(n):
    mylist = []
    for i in range(n):
        if i%2 == 0:
            mylist.append(3)
        elif i%2 == 1:
            mylist.append(8)
    return mylist

In [11]:
alt2(5)

[3, 8, 3, 8, 3]

In [12]:
alt2(6)

[3, 8, 3, 8, 3, 8]

(Using-NumPy)=
## Using NumPy

<iframe width="560" height="315" src="https://www.youtube.com/embed/wz2mRdkVmcI" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

The most important Python library for Math 9 is NumPy.  Because this is an external library, it needs to be installed on your computer before it can be used.  Once it is installed, it can be imported as in the following cell.  The convention (which we will always follow) is to abbreviate the library by `np`.

In [8]:
import numpy as np

Many parts of NumPy are very similar to Matlab.  For example, NumPy has a `zeros` function, just like Matlab.  We write `np.zeros` rather than just `zeros`, so that Python knows this function is defined in the NumPy library.

In [3]:
A = np.zeros((3,5))

This `A` is yet another type of object in Python.  It is a type defined in the NumPy library, called `ndarray`.  The "nd" portion stands for "n-dimensional".  In this case, `A` is two dimensional (it has a dimension of rows and a dimension of columns).

In [4]:
type(A)

numpy.ndarray

In [5]:
A

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [6]:
A.shape

(3, 5)

We are going to make a new version of our alternating function, this time using NumPy.  As a first step, here is a function that returns a length `n` NumPy array of all zeros.

In [7]:
def alt3(n):
    A = np.zeros(n)
    return A

In [9]:
alt3(7)

array([0., 0., 0., 0., 0., 0., 0.])

In [10]:
alt3(7)[0]

0.0

The decimal points are a hint that these zeros are floats rather than integers.  Just for practice with data types, let's specify that `A` should hold integers, not floats.

In [11]:
type(alt3(7)[0])

numpy.float64

In [12]:
A = alt3(7)

In [13]:
A.dtype

dtype('float64')

To change that data type from floats to integers, we will use the `dtype` argument.

In [15]:
help(np.zeros)

Help on built-in function zeros in module numpy:

zeros(...)
    zeros(shape, dtype=float, order='C', *, like=None)
    
    Return a new array of given shape and type, filled with zeros.
    
    Parameters
    ----------
    shape : int or tuple of ints
        Shape of the new array, e.g., ``(2, 3)`` or ``2``.
    dtype : data-type, optional
        The desired data-type for the array, e.g., `numpy.int8`.  Default is
        `numpy.float64`.
    order : {'C', 'F'}, optional, default: 'C'
        Whether to store multi-dimensional data in row-major
        (C-style) or column-major (Fortran-style) order in
        memory.
    like : array_like
        Reference object to allow the creation of arrays which are not
        NumPy arrays. If an array-like passed in as ``like`` supports
        the ``__array_function__`` protocol, the result will be defined
        by it. In this case, it ensures the creation of an array object
        compatible with that passed in via this argument.
   

Here we specify that the `dtype` should be 64-bit integers.

In [16]:
def alt3(n):
    A = np.zeros(n, dtype=np.int64)
    return A

In [17]:
alt3(8)

array([0, 0, 0, 0, 0, 0, 0, 0])

In [19]:
A = alt3(8)

In [20]:
A

array([0, 0, 0, 0, 0, 0, 0, 0])

NumPy supports many types of elementwise operations, just like Matlab.

In [21]:
A+3

array([3, 3, 3, 3, 3, 3, 3, 3])

Most standard types defined in Python do not support this sort of elementwise operation.  For example, `list` does not.

In [22]:
mylist = [0,0,0]

In [23]:
mylist+3

TypeError: can only concatenate list (not "int") to list

In [24]:
def alt3(n):
    A = np.zeros(n, dtype=np.int64)+3
    return A

In [26]:
A = alt3(5)

We now have a version of the alternating function that returns a NumPy array of all 3s.

In [27]:
A

array([3, 3, 3, 3, 3])

In [28]:
A[1::2]

array([3, 3])

Here is a convenient syntax (again, it does not work for lists) for setting the odd-indexed elements of `A` to be 8.

In [ ]:
A[1::2] = 8

In [30]:
A

array([3, 8, 3, 8, 3])

In [6]:
def alt3(n):
    A = np.zeros(n, dtype=np.int64)+3
    A[1::2] = 8
    return A

It's always a good idea, when writing a function like this, to test it on both an even input and on an odd input.

In [32]:
alt3(5)

array([3, 8, 3, 8, 3])

In [33]:
alt3(6)

array([3, 8, 3, 8, 3, 8])

In [34]:
mylist = [3,1,4,1,5,9]

In [35]:
mylist[::2]

[3, 4, 5]

In [36]:
mylist[::2] = -17

TypeError: must assign iterable to extended slice

Here we convert `mylist` to a NumPy array.

In [37]:
myarray = np.array(mylist)

Now, because `myarray` is a NumPy array, we can use the slice syntax to set the even-indexed entries to be equal to -17.

In [38]:
myarray[::2] = -17

In [39]:
myarray

array([-17,   1, -17,   1, -17,   9])

If you really want to make this assignment using `mylist`, you need the right-hand side to be the same length as the number of slots you are trying to fill in.  (So the NumPy version is much easier to use.)

In [40]:
mylist[::2] = [-17, -17, -17]

In [41]:
mylist

[-17, 1, -17, 1, -17, 9]

If you use the wrong number of terms on the right side, you will get an error.

In [42]:
mylist[::2] = [-17, -17]

ValueError: attempt to assign sequence of size 2 to extended slice of size 3

In [43]:
mylist[::2] = [-17, -17, -17, -17]

ValueError: attempt to assign sequence of size 4 to extended slice of size 3

## Timing different strategies

<iframe width="560" height="315" src="https://www.youtube.com/embed/BM20r8_BcGQ" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Here I have restarted this Jupyter notebook.  Let's compare some of the functions we have defined.

In [1]:
def alt3(n):
    A = np.zeros(n, dtype=np.int64)+3
    A[1::2] = 8
    return A

We need to remember to re-import NumPy each time we restart the notebook, otherwise we'll get an error like in the following.

In [2]:
alt3(10)

NameError: name 'np' is not defined

In [3]:
import numpy as np

In [4]:
def alt3(n):
    A = np.zeros(n, dtype=np.int64)+3
    A[1::2] = 8
    return A

In [5]:
alt3(10)

array([3, 8, 3, 8, 3, 8, 3, 8, 3, 8])

Here is another version, which is very similar to our original function from the top of this notebook (the only real difference is that I'm not using a separate `const` function in this case).

In [6]:
def alt4(n):
    mylist = []
    for i in range(n):
        mylist.append(3)
    for i in range(1,n,2):
        mylist[i] = 8
    return mylist

In [7]:
alt4(10)

[3, 8, 3, 8, 3, 8, 3, 8, 3, 8]

The functions `alt3` and `alt4` basically produce the same output, but `alt3` returns a NumPy array whereas `alt4` returns a list.

Let's evaluate how long `alt4` takes in the case of input `10`.  Here is an example of the unit of time being µs, which stands for microsecond, which is $10^{-6}$ seconds.

In [8]:
%%timeit
alt4(10)

1.18 µs ± 21.7 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


We have already used for loops and if statements.  A third type of fundamental programming construction is a while loop.  Just for practice with while loops (this will be our first time seeing them), let's try to find a value of `n` such that `alt4(n)` takes over 5 seconds to complete.  I don't know how to do this using the `%%timeit` Jupyter magic function, so instead, let's use another Python library, the time library.  This is a standard library in Python, so you don't need to install it.

In [9]:
import time

This number is not very meaningful by itself.  It becomes meaningful if we subtract another time from it, in which case we will get the number of seconds elapsed between the two calls.

In [10]:
time.time()

1657444015.338846

In [11]:
time.time()

1657444029.606127

In [12]:
start = time.time()
alt4(10**3)
end = time.time()
t = end-start

In [13]:
t

0.0003650188446044922

Here we keep multiplying `n` by 2 and checking the new time.  We will exit the while loop once we have a value of `n` which takes over 5 seconds to complete.  Try going through the following code, step-by-step, to convince yourself that this really does produce the first value of `n` we've found for which `alt4(n)` takes over 5 seconds to complete.

In [14]:
t = 0
n = 100
while t < 5:
    n = n*2 # n *= 2
    start = time.time()
    alt4(n)
    end = time.time()
    t = end-start

Here we check that the time really is bigger than 5.

In [15]:
t

5.250777959823608

The following value of `n` is the length we were looking for.  So it took about 5.25 seconds to produce a length 52,428,800  list, alternating between 3s and 8s.

In [16]:
n

52428800

In [17]:
def alt3(n):
    A = np.zeros(n, dtype=np.int64)+3
    A[1::2] = 8
    return A

Notice how much faster it is using the NumPy version of our function.  For the same input `n`, the `alt3` function takes about a quarter of a second instead of over 5 seconds.

In [18]:
print(n)
start = time.time()
alt3(n)
end = time.time()
t = end-start
print(t)

52428800
0.26982712745666504


If you want the output to be a list instead of a NumPy array, you can convert to a list.

In [19]:
def alt3b(n):
    A = np.zeros(n, dtype=np.int64)+3
    A[1::2] = 8
    return list(A)

In [20]:
alt3b(10)

[3, 8, 3, 8, 3, 8, 3, 8, 3, 8]

Surprisingly (to me), making this conversion to a list really slows down the function.

In [21]:
print(n)
start = time.time()
alt3b(n)
end = time.time()
t = end-start
print(t)

52428800
4.124554872512817


Here is a mix between the two methods.  It uses a for loop to assign the 8 values, instead of using the slicing syntax.  The main lesson is that, when possible, we should always try to avoid for loops.  The assignment using `A[1::2] = 8` was much faster, over 10 times faster, than in the following where we make the same assignment using a for loop.  

In [22]:
def alt3c(n):
    A = np.zeros(n, dtype=np.int64)+3
    for i in range(1,n,2):
        A[i] = 8
    return A

In [23]:
print(n)
start = time.time()
alt3c(n)
end = time.time()
t = end-start
print(t)

52428800
2.886206865310669
